### PR TITLE
BatchNorm channels last memory format optimization on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
BatchNorm channels last memory format optimization on CPU

### CODE REVIEW

**Comment by mingfeima** [comment]
* BatchNorm training forward path is paralleled but inference path is not paralleled at the moment. This explains why inference path is slower than training forward on multiple cores in `batchnorm_M1_N256_K3136_cpu_bwdall `.
* Moving channels last impl to /cpu doesn't bring any performance improvement (gcc should be able to vectorize simple a * x + b pattern pretty well) but i still did so since the overall code structure will be cleaner. For example, you can keep just one version of `batch_norm_cpu_inference_collect_linear_and_constant_terms`.
* I noticed the original impl has a special treatment for `image_size=1`, it is vectorized in this patch, so for K1 cases there are some improvement.
* Since the vectorization pattern is separately treated for `image_size=1` and `channels=1`, I add new tests to cover these scenarios.


Update performance with [pytorch internal operator benchmark](https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark), machine Xeon(R) Gold 6248 CPU, 20 cores per socket, 2.5GHz. **1C** refers to single core run, **20C** refers to single socket run. jemalloc and numactrl are applied to reduce test result fluctuation.


Name | Tag | before (1C) | before (20C) | after (1C) | after (20C)
-- | -- | -- | -- | -- | --
batchnorm_M1_N256_K3136_cpu | Short | 248.093 | 252.302 | 255.109 | 22.736
batchnorm_M1_N256_K3136_cpu_bwdall | Short | 3605.876 | 404.381 | 3620.332 | 401.902
batchnorm_M1_N256_K3136_cpu_bwd1 | Short | 3600.249 | 403.502 | 3573.064 | 403.201
batchnorm_M1_N8192_K1_cpu | Long | 26.128 | 25.874 | 24.17 | 23.457
batchnorm_M1_N2048_K1_cpu | Long | 10.377 | 10.455 | 9.845 | 9.978
batchnorm_M128_N8192_K1_cpu | Long | 665.283 | 670.573 | 375.329 | 53.614
batchnorm_M128_N2048_K1_cpu | Long | 177.345 | 184.846 | 98.43 | 22.127
batchnorm_M1_N8192_K1_cpu_bwdall | Long | 16337.803 | 5278.381 | 15562.81 | 5200.735
batchnorm_M1_N8192_K1_cpu_bwd1 | Long | 16343.511 | 5712.935 | 15526.72 | 4797.94
batchnorm_M1_N2048_K1_cpu_bwdall | Long | 4098.062 | 1397.278 | 3894.704 | 1289.964
batchnorm_M1_N2048_K1_cpu_bwd1 | Long | 4088.836 | 1297.217 | 3902.353 | 1390.161
batchnorm_M128_N8192_K1_cpu_bwdall | Long | 36745.255 | 5309.559 | 34362.76 | 5134.023
batchnorm_M128_N8192_K1_cpu_bwd1 | Long | 36715.947 | 5316.02 | 34458 | 5102.021
batchnorm_M128_N2048_K1_cpu_bwdall | Long | 7132.537 | 1307.126 | 7074.514 | 1284.771
batchnorm_M128_N2048_K1_cpu_bwd1 | Long | 7065.365 | 1361.857 | 6962.995 | 1343.29

**Comment by VitalyFedyunin** [comment]
Code looks good, but I see that you changed channels first AND channels last implementations. Can you please make sure to benchmark both memory layouts.

**Review by VitalyFedyunin — APPROVED** [review]
I will approve to continue process, but I will not land until benchmarks provided.

**Comment by mingfeima** [comment]
On the channels last format, inference performance is OK but training is not good enough since the current [impl](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L188) on training will parallel on dim of `channels` which leaves each `NHW` plain on nhwc format non-contiguous.

I need to continue to improve this PR, once finished I will stack this with other channels last pull requests.

**Comment by mingfeima** [comment]
please review the new one #48919
